### PR TITLE
In PR 

### PR DESCRIPTION
**Pull request by Chenyu-Shi**
finishing in expression. adding more tests and null support. need confirmation on null behavior and also I wonder why integer field is sufficient for string comparison. will bring this up at the meeting.

### CODE REVIEW

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 1]
```diff
@@ -702,6 +702,92 @@ class FlatbuffersExpressionEvaluator {
...
+      size_t num_children = c->children()->size(); 
+      bool result = false;
+
+      if (num_children < 2){
+        throw std::runtime_error(std::string("In can't operate with fewer than 2 args, currently we have")
+          + std::to_string(num_children)); 
+      }
+
+      auto left_offset =  eval_helper(row, (*c->children())[0]);
+      const tuix::Field *left = flatbuffers::GetTemporaryPointer(builder, left_offset);
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:717]: We need to be able to handle `NULL` values, an [example](https://github.com/mc2-project/opaque/blob/master/src/enclave/Enclave/ExpressionEvaluation.h#L40) here.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 2]
```diff
@@ -702,6 +702,92 @@ class FlatbuffersExpressionEvaluator {
       }
     }
 
+    case tuix::ExprUnion_In:
+    {
+      auto c = static_cast<const tuix::In *>(expr->expr());
+      size_t num_children = c->children()->size(); 
+      bool result = false;
+
+      if (num_children < 2){
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:711]: Can you verify what kind of types `IN` handles? What if the user passes in a CalendarType?

**Review by wzheng — COMMENTED** [review]
Partial review.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 3]
```diff
@@ -702,6 +702,92 @@ class FlatbuffersExpressionEvaluator {
...
+        const tuix::Field *item = flatbuffers::GetTemporaryPointer(builder, right_offset);
+        if (item->value_type() != left->value_type()){
+          throw std::runtime_error(
+          std::string("In can't operate on ")
+          + std::string(tuix::EnumNameFieldUnion(left->value_type()))
+          + std::string(" and ")
+          + std::string(tuix::EnumNameFieldUnion(item->value_type())));
+        }
+        //TODO why integer field passing the string test 
+        if (static_cast<const tuix::IntegerField *>(item->value())->value() == static_cast<const tuix::IntegerField *>(left->value())->value()){
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h]: Used the dynamically comparison operator to fix this problem.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 4]
```diff
@@ -702,6 +702,92 @@ class FlatbuffersExpressionEvaluator {
...
+      }
+      return tuix::CreateField(
+        builder,
+        tuix::FieldUnion_BooleanField,
+        tuix::CreateBooleanField(builder, result).Union(),
+        left->is_null());
+
+    }
+
+    case tuix::ExprUnion_Concat:
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h:758]: sorry this is still in the same PR but both are somewhat complete now. I will look into splitting PR in the future.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 5]
```diff
@@ -702,6 +702,92 @@ class FlatbuffersExpressionEvaluator {
...
+      /*
+      auto array_field = static_cast<const tuix::ArrayField *>(value->value());
+      std::string str = to_string(array_field);
+      std::vector<uint8_t> str_vec(str.begin(), str.end());
+        return tuix::CreateField(
+          builder,
+          tuix::FieldUnion_StringField,
+          tuix::CreateStringFieldDirect(builder, &str_vec, str_vec.size()).Union(),
+          result_is_null);
+          */
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h]: fixed

File: src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala [hunk 6]
```diff
@@ -965,6 +967,20 @@ object Utils extends Logging {
...
+            tuix.In.createIn(
+              builder, tuix.In.createChildrenVector(builder, childrenOffsets.toArray)))
+
+        case (Concat(child), childrenOffsets) =>
+          tuix.Expr.createExpr(
+            builder,
+            tuix.ExprUnion.Concat,
+            tuix.Concat.createConcat(
+              builder, tuix.Concat.createChildrenVector(builder, childrenOffsets.toArray)))
+
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala:983]: sorry this is still in the same PR but both are somewhat complete now. I will look into splitting PR in the future.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 7]
```diff
@@ -702,6 +702,92 @@ class FlatbuffersExpressionEvaluator {
...
     }
 
+    case tuix::ExprUnion_In:
+    {
+      auto c = static_cast<const tuix::In *>(expr->expr());
+      size_t num_children = c->children()->size(); 
+      bool result = false;
+
+      if (num_children < 2){
+        throw std::runtime_error(std::string("In can't operate with fewer than 2 args, currently we have")
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h]: fixed

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 1 above)
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h:717]: fixed

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 2 above)
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h:710]: Used a dynamic typed comparison method that should work with different types

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 8]
```diff
@@ -702,6 +702,102 @@ class FlatbuffersExpressionEvaluator {
...
+        // adding dynamic casting 
+        bool temporary_result =
+        static_cast<const tuix::BooleanField *>(
+          flatbuffers::GetTemporaryPointer<tuix::Field>(
+            builder,
+            eval_binary_comparison<tuix::EqualTo, std::equal_to>(
+              builder,
+              flatbuffers::GetTemporaryPointer<tuix::Field>(builder, left_offset),
+              flatbuffers::GetTemporaryPointer<tuix::Field>(builder, right_offset)))
+          ->value())->value();
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:743]: This doesn't work for every type, should you check for the supported type here?

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 9]
```diff
@@ -389,6 +389,59 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
     df.filter($"word".contains(lit("1"))).collect
   }
 
+  testAgainstSpark("concatwithstring") { securityLevel =>
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala]: fixed

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 10]
```diff
@@ -389,6 +389,59 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
...
+    df.select(concat(col("str"),lit(","),col("int"))).collect
+  }
+
+  testAgainstSpark("isin1") { securityLevel =>
+    //normal test 1 
+    val ids = Seq((1, 2, 2), (2, 3, 1))
+    val df = makeDF(ids, securityLevel, "x", "y", "id")
+    val c = $"id" isin ($"x", $"y")
+    val result = df.filter(c)
+    //result.explain(true)
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala]: fixed

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 11]
```diff
@@ -389,6 +389,59 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
...
+    result.collect 
+  }
+  testAgainstSpark("isin2") { securityLevel =>
+    //normal test 2     
+    val ids2 = Seq((1, 1, 1), (2, 2, 2), (3,3,3), (4,4,4))
+    val df2 = makeDF(ids2, securityLevel, "x", "y", "id")
+    val c2 = $"id" isin (1 ,2, 4, 5, 6)
+    val result = df2.filter(c2)
+    //result.explain(true)
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala]: fixed

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 12]
```diff
@@ -389,6 +389,59 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
...
+    result.collect 
+  }
+  testAgainstSpark("isin3") { securityLevel =>
+    //string test 
+    val ids3 = Seq(("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"), ("b", "b", "b"), ("c","c","c"), ("d","d","d"))
+    val df3 = makeDF(ids3, securityLevel, "x", "y", "id")
+    val c3 = $"id" isin ("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" ,"b", "c", "d", "e")
+    val result = df3.filter(c3)
+    //result.explain(true)
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala]: fixed

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 13]
```diff
@@ -389,6 +389,59 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
...
+    result.collect
+  }
+  testAgainstSpark("isin4") { securityLevel =>
+    //null test  
+    val ids4 = Seq((1, 1, 1), (2, 2, 2), (3,3,null.asInstanceOf[Int]), (4,4,4))
+    val df4 = makeDF(ids4, securityLevel, "x", "y", "id")
+    val c4 = $"id" isin (null.asInstanceOf[Int])
+    val result = df4.filter(c4)
+    //result.explain(true)
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala]: fixed

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 14]
```diff
@@ -702,6 +702,102 @@ class FlatbuffersExpressionEvaluator {
...
+      for (size_t i =0; i< num_children; i++){
+              auto offset =  eval_helper(row, (*c->children())[i]);
+              const tuix::Field *str = flatbuffers::GetTemporaryPointer(builder, offset);
+              if (str->value_type() != tuix::FieldUnion_StringField) {
+                throw std::runtime_error(
+                  std::string("tuix::Concat requires string, not ")
+                  + std::string(tuix::EnumNameFieldUnion(str->value_type()))
+                  + std::string(". You do not need to provide the data as string but the data should be serialized into string before sent to concat"));  
+              }
+              if (!str->is_null()){
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h:777]: The final result is not NULL if one value is null. It just skips over the null value. Modified a test to reflect this

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 8 above)
Comment (Chenyu-Shi) [reply src/enclave/Enclave/ExpressionEvaluation.h:743]: The eval_binary_comparison does the type checking and throws an type error if incompatible data type is passed in which is sufficient for type checking?

**Comment by wzheng** [comment]
Closing this PR since this has been split into #124 and #125.
